API Change Proposal: Events Loop 2.0

# UPDATE 8/23/18:
A PR for Windows is available at #638

# UPDATE 7/12/18:
I've adapted the agreed-upon changes into a changelog here: https://gist.github.com/Osspial/d3a3390807fa95a469d9c9b677db4715. The original proposal is below.

------

# Events Loop 2.0

This issue proposes a series of changes to winit's `EventsLoop` API, based around the experiences I've had with creating a GUI library that uses winit as a window creation backend. The main goal is to give winit complete control over the event loop, both for event polling and event waiting. These changes should let us fix various implementation bugs on the win32 backend, as well as fix what I believe are design bugs in the current API.

## Technical Motivation:

**EDIT:** The technical motivation for the Windows backend is included below. However, pretty much every platform winit supports is hacky or broken in some way by necessity of the current API we expose.

Right now, the `win32` backend creates a background thread, in which the events loop resides. The biggest problem with this is that it introduces various unexpected synchronization bugs that require a good amount of effort to fix and increase the complexity of the backend (see #415 and #391). However, it's necessary to cleanly handle `win32`'s various modal event loops.

So, what are these modal loops, and why do they matter? On Windows, when a user begins resizing a window, Windows suspends the program's main event loop and begins its own loop, inside of the resize event's handler. Unfortunately, this freezes any processing that's going on outside the Windows event loop, which is currently worked around by having the loop exist in its own thread, introducing quite a few bugs as mentioned above. Inside the loop, Windows does a bunch of additional processing, most notably handling window snapping (although it probably does other stuff too that isn't obvious).

Windows doesn't publicly expose the APIs used to handle snapping, so re-implementing the modal loops ourselves so that they don't suspend the main event loop isn't really an option. However, Windows should still process events we push to the Windows events queue, letting us implement event polling inside the Windows event loop with custom events ([using the same mechanisms used for wakeup messages](https://github.com/tomaka/winit/blob/bdc01fee1a54470b07a897592927c26d1af60341/src/platform/windows/events_loop.rs#L312-L328)), as long as winit's API lets it assume complete control over program control flow. Right now, it doesn't, but this change would let us remove the second thread and simplify much of the implementation.

Actually removing the second thread isn't that difficult (see #445), but these bugs present pretty major blockers that should be resolved before that's merged.


## Design Details

**EDIT:** The original design has been removed from this post, as it is outdated and may cause confusion for new readers. However, it does contain some useful information about the rationale behind the changes, and can be accessed here: https://gist.github.com/Osspial/53efa87797899872650f6bece14736c0

@Osspial Yeah, this new API change has the same problem that I feared it would have. Take this code:

```rust
extern crate winit;

use winit::window::WindowBuilder;
use winit::platform::desktop::EventLoopExtDesktop;
use winit::event::{Event, WindowEvent};
use winit::event_loop::{EventLoop, ControlFlow};

fn main() {

    #[derive(Debug)]
    pub struct Data { internal: f32 };

    let mut data = Data { internal: 5.0 };

    let mut events_loop = EventLoop::new();

    let _window = WindowBuilder::new()
        .with_title("A fantastic window!")
        .build(&events_loop)
        .unwrap();

    events_loop.run_return(move |event, _event_loop, control_flow| {
        match event {
            Event::WindowEvent { event, .. } => {
                match event {
                    WindowEvent::CloseRequested => {
                        *control_flow = ControlFlow::Exit;
                    },
                    e => {
                        data.internal += 1.0;
                        *control_flow = ControlFlow::Wait;
                    },
                }
            },
            _ => *control_flow = ControlFlow::Wait,
        }
    });

    println!("Hello from main, data: {:?}!", data);
}
```

This doesn't compile:

```rust
error[E0382]: use of moved value: `data`
  --> examples\window.rs:39:46
   |
21 |     events_loop.run_return(move |event, _event_loop, control_flow| {
   |                            --------------------------------------- value moved (into closure) here
...
39 |     println!("Hello from main, data: {:?}!", data);
   |                                              ^^^^ value used here after move
   |
   = note: move occurs because `data` has type `main::Data`, which does not implement the `Copy` trait

error: aborting due to previous error
```

Essentially, I have some application data, and I do want to use my application data after the function returns. But I can't do that, even with `Rc` and `clone`, because **everything** is moved into the closure:

```rust
extern crate winit;

use winit::window::WindowBuilder;
use winit::platform::desktop::EventLoopExtDesktop;
use winit::event::{Event, WindowEvent};
use winit::event_loop::{EventLoop, ControlFlow};

fn main() {
    use std::rc::Rc;
    use std::cell::RefCell;

    #[derive(Debug)]
    pub struct Data { internal: f32 };

    let data = Data { internal: 5.0 };
    let data_wrapper = Rc::new(RefCell::new(data));

    let mut events_loop = EventLoop::new();

    let _window = WindowBuilder::new()
        .with_title("A fantastic window!")
        .build(&events_loop)
        .unwrap();

    let mut data_return = None;

    events_loop.run_return(move |event, _event_loop, control_flow| {
        println!("data: {:?}", data_wrapper);
        match event {
            Event::WindowEvent { event, .. } => {
                match event {
                    WindowEvent::CloseRequested => {
                        *control_flow = ControlFlow::Exit;
                        data_return = Some(data_wrapper.clone());
                    },
                    _ => {
                        data_wrapper.borrow_mut().internal += 1.0;
                        *control_flow = ControlFlow::Wait;
                    },
                }
            },
            _ => *control_flow = ControlFlow::Wait,
        }
    });

    println!("Hello from main, data: {:?}!", data_return);
}
```

```rust
error[E0382]: use of moved value: `data_return`
  --> examples\window.rs:46:46
   |
27 |     events_loop.run_return(move |event, _event_loop, control_flow| {
   |                            --------------------------------------- value moved (into closure) here
...
46 |     println!("Hello from main, data: {:?}!", data_return);
   |                                              ^^^^^^^^^^^ value used here after move
   |
   = note: move occurs because `data_return` has type `std::option::Option<std::rc::Rc<std::cell::RefCell<main::Data>>>`, which does not implement the `Copy` trait

error: aborting due to previous error
```

Would it be possible for the `run_return` function (the `run` function doesn't need this necessarily, as it exits the program on return) to accept a fourth parameter, to pass some &T to the callback? This way, I could "smuggle" the data in that `T`, so that it doesn't have to be moved into the callback. I'm not sure if this is possible however, given the constraints of Win32 APIs.

Now, this isn't a total dealbreaker, as many applications don't require that style of programming, but I do want to higlight it. On the other hand, the `move || ` has very nicely solved the problem with Windows callbacks - so that the final application only requires one thread instead of 2, so I can do a `SetMenu` call without it crashing immediately. Really nice job on that.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,7 @@
 # Unreleased
 
+# 0.20.0 Alpha 1
+
 - Changes below are considered **breaking**.
 - Change all occurrences of `EventsLoop` to `EventLoop`.
 - Previously flat API is now exposed through `event`, `event_loop`, `monitor`, and `window` modules.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "winit"
-version = "0.19.1"
+version = "0.20.0-alpha1"
 authors = ["The winit contributors", "Pierre Krieger <pierre.krieger1708@gmail.com>"]
 description = "Cross-platform window creation library."
 keywords = ["windowing"]

--- a/README.md
+++ b/README.md
@@ -7,7 +7,7 @@
 
 ```toml
 [dependencies]
-winit = "0.19.1"
+winit = "0.20.0-alpha1"
 ```
 
 ## [Documentation](https://docs.rs/winit)

